server/svix-server/src/worker.rs

```diff
@@ -994,13 +994,13 @@ pub async fn queue_handler(
                             .await
                             .is_err()
                         {
-                            if let Err(err) = queue_tx.nack(delivery).await {
+                            if let Err(err) = delivery.nack().await {
```

Comment:
Unfortunately, we can't add `ack`/`nack` methods to the `TaskQueueReceiver`. From the worker
```rust
match queue_rx.receive_all().await {
            Ok(batch) => {
                for delivery in batch {
                    // ...
                    tokio::spawn(async move {
                        let worker_context = WorkerContext { ... };
                        let queue_task = // ...
                        if process_queue_task(worker_context, queue_task)
                            .await
                            .is_err()
                        {
                            if let Err(err) = delivery.nack().await {
                                // ...
                            }
                        } else if let Err(err) = delivery.ack().await {
                            // ...
                        }
                    });
                }
            }
            // ...
        }
```
`queue_rx.receive_all()` takes `queue_rx` by a mutable reference, so we can't have a reference to it in the spawned task ☹️

Reply:
You can use `RefCell` to workaround that and have inner mutability. I think that makes more sense than changing the whole API. Especially since I think it's nicer having the operations on the queue rather than on the "delivery".

What do you think?

Reply:
Okay so my explanation was somewhat incomplete before.

There are two issues at play. The 1st is the mutability of `receive_all()`, which prevents `ack()` and `nack()` being called inside the loop. The 2nd is that `tokio::spawn(async move { ... })` necessarily takes ownership of everything passed in. Even if `receive_all()` did take `&self` (instead of `&mut self`), we'd have to `clone()` the `queue_rx`.

`Arc<RefCell<...>>` won't work here, since `RefCell` isn't sync, we won't be able to pass it into the spawned task (we'd get an error like `future cannot be sent between threads safely
the trait Sync is not implemented for queue_rx`). We could use `Mutex`, but now there's an unnecessary runtime cost to `queue_rx.receive_all()`.

So I agree in theory, I'd rather have `queue_rx.ack(delivery)` than `delivery.ack()`. But I think the tradeoff isn't worth it